### PR TITLE
[Enhancement] improve full statistics collect strategy

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

1. The algorithm for statistics's health has been modified:
Before: `health = 1 - updated rows / minimum partition rows`
After: when the number of updated partitions is small, use `1 - updated rows / rows in the updated partition` ; when the number of updated partitions is large, use `1 - max(updated rows / rows in the updated partition, updated partitions / total partitions)`
 

2. Add config for schedule process
statistic_auto_sample_ratio
statistic_auto_sample_data_size
statistic_auto_collect_interval

now logical:

1. Check statistics has collected after table update?
2. Check the last collect job is greater than `statistic_auto_collect_interval` interval.
3. Check health:
  a. health > statistic_auto_collect_ratio:   small rows has updated in the table, and the statistics isn't much different, so there is no need to collect again
  b. health > statistic_auto_sample_ratio && health < statistic_auto_collect_ratio: some rows has updated in the table,  we should updated statistics
  c. health < statistic_auto_sample_ratio: a large rows has updated in the table, the table was almost rewrite in `statistic_auto_collect_interval` time, may is a real-time update table, we need check more condition:
    i). if table size is small (<statistic_auto_sample_data_size): we believe full statistics has little impact, do full collect 
    ii). if table size is large (>statistic_auto_sample_data_size): we will do sample collect

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/statistic/StatisticsCollectJobFactory.java [hunk 1]
```diff
@@ -115,32 +121,63 @@ private static void createJob(List<StatisticsCollectJob> allTableJobMap, Analyze
...
+                    LOG.debug("statistics job choose sample on real-time update table: {}" +
+                                    ", last collect time: {}, current healthy: {}, full collect healthy limit: {}<, " +
+                                    ", update data size: {}MB, full collect healthy data size limit: <{}MB",
+                            table.getName(), statisticsUpdateTime, healthy,
+                            Config.statistic_auto_sample_ratio, ByteSizeUnit.BYTES.toMB(sumDataSize),
+                            ByteSizeUnit.BYTES.toMB(Config.statistic_auto_sample_data_size));
+                    allTableJobMap.add(buildStatisticsCollectJob(db, table, null, columns,
+                            StatsConstants.AnalyzeType.SAMPLE, job.getScheduleType(), job.getProperties()));
+                    return;
+                }
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/statistic/StatisticsCollectJobFactory.java:180]: Is this logic used to auto update the type of auto collect job?  Maybe add some comments or extract this to a new method?

File: fe/fe-core/src/main/java/com/starrocks/statistic/BasicStatsMeta.java [hunk 2]
```diff
@@ -110,40 +110,44 @@ public Map<String, String> getProperties() {
...
         for (Partition partition : table.getPartitions()) {
             if (!partition.hasData()) {
-                //skip init empty partition
+                // skip init empty partition
                 continue;
             }
-            if (partition.getRowCount() < minRowCount) {
-                minRowCount = partition.getRowCount();
+
+            LocalDateTime.class.getClassLoader();
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/statistic/BasicStatsMeta.java:122]: sorry, will remove

File: fe/fe-core/src/main/java/com/starrocks/statistic/SampleStatisticsCollectJob.java [hunk 3]
```diff
@@ -79,6 +94,19 @@ public void collect(ConnectContext context, AnalyzeStatus analyzeStatus) throws
...
     }
 
+    private String getDataSize(Column column) {
+        if (column.getPrimitiveType().isCharFamily()) {
+            return "IFNULL(SUM(CHAR_LENGTH(`column_key`)), 0)";
+        }
+
+        long typeSize = column.getType().getTypeSize();
+
+        if (column.getType().canStatistic()) {
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/statistic/SampleStatisticsCollectJob.java:104]: need `getSize` logic, but it's different with sample statistics

File: fe/fe-core/src/main/java/com/starrocks/statistic/StatsConstants.java [hunk 4]
```diff
@@ -23,6 +23,7 @@ public class StatsConstants {
     public static final int STATISTIC_HISTOGRAM_VERSION = 2;
     public static final int STATISTIC_TABLE_VERSION = 3;
 
+    public static final int STATISTICS_HEALTH_LAST_PARTITIONS = 10;
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/statistic/StatsConstants.java:26]: This isn't a good name. I want to express the different partition strategies when the health check, do you have any suggestions?

[the remark this answers is not in the document]
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/statistic/StatsConstants.java:26]: how about `STATISTICS_PARTITION_UPDATED_THRESHOLD` ?

[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/statistic/StatsConstants.java:26]: nice！

**Review by packy92 — DISMISSED** [review]

**Review by liuyehcf — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**